GDScript: Fix constant conversions

vonagam opened the pull request:
A constant value can be converted to a specified type during analysis, not at runtime every time.

```gdscript
var x: StringName = 'a' # initializer

x = 'b' # assignment

func print_name(y: StringName): print(y)
print_name('c') # call argument

func make_name() -> StringName: return 'd' # return

var z := 'c' as StringName # cast
```

Out of those places only variable initializer was handled at compile time before. 

Initializer for constants was not converted at all which was leading to many crash reports.

While at it, allowed a cast to be considered as constant expression under certain circumstances (which will be expanded with array PR from which code for this PR partially comes from).

Fixed minor things with casts and arguments validations.

Fixes #70861.
Fixes #71255.
Fixes #71483.
Fixes #71635.
Fixes #72209.

Related to #70872 (supersedes).
Related to #70873 (supersedes).

vonagam committed:
GDScript: Fix constant conversions

dalexeev commented:
This PR does not fully supersede #70873, as there are more changes to `resolve_assignable()`, not just the part that can be replaced by `update_const_expression_builtin_type()` call.

vonagam commented:
Hm. So I assume that the thing that wasn't included here was a fix for incorrect check of asymmetric compatibility - String can be converted to Color, but other way around will never work. Yeah, wrong argument was passed to `is_type_compatible` (`true` instead of `false`) in reverse check (that should cover only objects). Updated the PR and fixed the same thing in assignments.

Was there anything else that is missing?

dalexeev commented on modules/gdscript/gdscript_analyzer.cpp:
```diff
@@ -2063,6 +2069,59 @@ void GDScriptAnalyzer::reduce_array(GDScriptParser::ArrayNode *p_array) {
+	if (p_expression->get_datatype() == p_type) {
+		return;
+	}
+	if (p_type.kind != GDScriptParser::DataType::BUILTIN && p_type.kind != GDScriptParser::DataType::ENUM) {
+		return;
+	}
```
I'm not sure if a silent return here is correct. It seems to me that either `ERR_FAIL_COND` is needed here, or the method should return `bool`.

vonagam replied:
This is not an error case, if an expected type is not built-in or enum (so a variant or an object) then there is nothing to do - no conversion will take place. There are checks afterwards in all corresponding places that deal with checking that types do match. 

Usage of the method assumes that you call it before storing the resulting type of an expression, like `update_array_literal_element_type` and `reduce_expression`, no need to return booleans.

dalexeev commented:
> Is there anything else that is missing?

1. Some tests are missing.
2. I think in my PR the code has been reorganized to be more readable: first the status is determined (`SAFE`, `UNSAFE`, `ERROR`) and then actions are performed depending on this. This can help avoid mistakes in the future.

dalexeev commented on modules/gdscript/gdscript_analyzer.cpp:
```diff
@@ -2063,6 +2069,59 @@ void GDScriptAnalyzer::reduce_array(GDScriptParser::ArrayNode *p_array) {
+
+	if (p_type.kind == GDScriptParser::DataType::ENUM && value_type.builtin_type == Variant::INT && enum_get_value_name(p_type, p_expression->reduced_value) == StringName()) {
+		if (expression_type.kind == GDScriptParser::DataType::ENUM) {
+			push_error(vformat(R"(Cannot %s "%s.%s" as Enum "%s": no corresponding enum member has the same value (%d).)", p_usage, expression_type.enum_type, enum_get_value_name(expression_type, p_expression->reduced_value), p_type.to_string(), (int64_t)p_expression->reduced_value), p_expression);
```
I think this should be a warning, not an error (so as not to create false expectations for the user), since in other cases it is not guaranteed that the value of a variable of type `MyEnum` will be from `MyEnum.values()`.

vonagam replied:
This case happens in casts when a value of one enum casted to represent another.
This is preexisting logic.
Don't understand why would you want it to be a warning instead of an error.
```gdscript
enum A { X = 0 }
enum B { Y = 1 }
var b := A.X as B
```

dalexeev replied:
There is no guarantee that a variable of type `MyEnum` cannot contain an arbitrary `int`. It is wrong that in some cases this is an error, and in others a warning.

```gdscript
enum MyEnum {
    VALUE_A,
}

func _run() -> void:
    var x: int = 42
    var a: MyEnum = x # Warning, not an error.
    print(a) # 42
```

vonagam replied:
In your case it will not apply `var x` is not a constant.

dalexeev replied:
![](https://user-images.githubusercontent.com/47700418/213986558-70d1f695-eeac-468f-8220-64a4f6218484.png)

vonagam replied:
`42` is a constant in that case. And there is no value 42 in `MyEnum`. How is it not an error?

dalexeev replied:
I wrote above. Enums are mostly an alias for `int`. An error **in this case** may give the user the impression that they cannot assign the wrong number **at all**. But this is not true in general.

vonagam replied:
Enum is a quasi-opaque type: an enum value is an int, but an int is not an enum. That's why there is a warning (when we don't know if there is a mismatch) and an error (when we do know that). Enum is not just an alias for int, it has its own type. 

Why type it `MyEnum` and not `int` if you want to assign random values to it?

dalexeev replied:
If the variable `a` **can be** equal to `42` (and this is true), then we should not forbid it in some particular cases, while in other cases there will be no error. We need to warn the user that they are doing something wrong (when we can detect it).

vonagam replied:
I mean `a` should not be ever equal to `42` in your example. I guess, there are missing checks in runtime to give an error when `42` is assigned to a variable of type `MyEnum`. I think it is just because enum information is stripped in runtime, maybe it was changed and it is possible to give an error now.

I recommend asking about what enum type represent in developer chat.

dalexeev replied:
See also [the comment](https://github.com/godotengine/godot/issues/71763#issuecomment-1399234932).

vonagam replied:
Ok, I asked in developer chat to clarify what does enum type represent.

But if you think that any int can be assigned to an enum type, then why do you think that there should be a warning then?

dalexeev replied:
Because it can give users false assumptions about what an enum is in GDScript. Errors should be only for those things that are not allowed in principle. If something is possible but wrong, then it should be a warning, not an error.

vonagam replied:
I mean if it is wrong then you make it an error to make it impossible. Why allow wrong code?

dalexeev replied:
I think we don't understand each other. I wrote above several times why I think so.

This is wrong code in the "shouldn't do this" sense, but not in the "can't do this" sense.

vonagam replied:
Done. It is a warning now, not an error.

akien-mga commented:
Needs rebase.

akien-mga commented:
Needs another rebase to pass CI, I fixed the GDScript test error.

akien-mga commented:
Thanks!
